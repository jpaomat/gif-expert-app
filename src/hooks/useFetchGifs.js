import { useEffect, useState } from "react"
import { getGifs } from "../services/getGifs";

export const useFetchGifs = (category) => {   //custom hook
    const [state, setstate] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
        getGifs(category).then(imgs => {
            setstate({
                data: imgs,
                loading: false
            })
        }); // como devuelve un solo argumento se puede llamar asÃ­
    }, [category]);

    return state;
}
