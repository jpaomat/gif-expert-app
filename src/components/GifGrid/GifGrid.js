import React, { useEffect, useState } from 'react';
import { useFetchGifs } from '../../hooks/useFetchGifs';
import { getGifs } from '../../services/getGifs';
import { GifGridItem } from '../GifGridItem/GifGridItem';
import PropTypes from 'prop-types';
import './GifGrid.css';

export const GifGrid = ({ category }) => {

    // PRIMER FORMA DE HACERLO
    // const [gifsList, setGifsList] = useState([]);
    // useEffect(() => {
    //     getGifs(category).then(setGifsList); // como devuelve un solo argumento se puede llamar as√≠
    // }, [category])

    // ALTERNATIVA PARA QUE QUEDE ORDENADO Y MAS LIMPIO EL COMPONENTE
    const {data:gifsList, loading} = useFetchGifs(category);

    return (
        <>
            <h3>{category}</h3>
            {loading && 'Cargando...'}
            <div className='card-grid'>
                {
                    gifsList.map(imgData => (
                        <GifGridItem
                            key={imgData.id}
                            {...imgData}
                        />
                    ))
                }
                <br></br>
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category : PropTypes.string.isRequired
}

