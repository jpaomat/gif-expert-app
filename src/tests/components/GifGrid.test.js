import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock("../../hooks/useFetchGifs");

describe('Test to GifGrid component', () => {
    const props = {
        category: 'cars'
    };
    let wrapper;

    test('should create component', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        wrapper = shallow(<GifGrid {...props} />);
        expect(wrapper).toMatchSnapshot();
    })

    test('should show items when images are loaded', () => {
        const gifs = [{
            id: '124',
            title: 'Toy Story',
            url: 'https://toystory.jpg'
        }];
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        wrapper = shallow(<GifGrid {...props} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })

})
