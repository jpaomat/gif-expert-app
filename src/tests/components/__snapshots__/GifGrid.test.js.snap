// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to GifGrid component should create component 1`] = `
<Fragment>
  <h3>
    cars
  </h3>
  Cargando...
  <div
    className="card-grid"
  >
    <br />
  </div>
</Fragment>
`;

exports[`Test to GifGrid component should show items when images are loaded 1`] = `
<Fragment>
  <h3>
    cars
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="124"
      key="124"
      title="Toy Story"
      url="https://toystory.jpg"
    />
    <br />
  </div>
</Fragment>
`;
